File: java-workshop/part-1-powermock/src/main/java/domain/Order.java
package domain;

import java.util.List;
public record Order(
    String id,
    String customerId,
    List<OrderItem> items,
    OrderStatus status,
    double totalPrice,
    String transactionId,
    String trackingNumber
) {
    public Order {
        items = List.copyOf(items);
    }

    public Order confirm(double totalPrice, String transactionId, String trackingNumber) {
        return new Order(id, customerId, items, OrderStatus.CONFIRMED, totalPrice, transactionId, trackingNumber);
    }

    public Order reject() {
        return new Order(id, customerId, items, OrderStatus.REJECTED, totalPrice, transactionId, trackingNumber);
    }

    public Order withStatus(OrderStatus status) {
        return new Order(id, customerId, items, status, totalPrice, transactionId, trackingNumber);
    }
}

File: csharp-workshop/Part1.StaticDeps/Domain/Order.cs
namespace Part1.StaticDeps.Domain;

public record Order(
    string Id,
    string CustomerId,
    IReadOnlyList<OrderItem> Items,
    OrderStatus Status,
    double TotalPrice,
    string? TransactionId,
    string? TrackingNumber)
{
    // Defensive copy — mirrors Java's compact constructor: public Order { items = List.copyOf(items); }
    public IReadOnlyList<OrderItem> Items { get; init; } = Items.ToList().AsReadOnly();

    public Order Confirm(double totalPrice, string transactionId, string trackingNumber) =>
        this with
        {
            Status = OrderStatus.Confirmed,
            TotalPrice = totalPrice,
            TransactionId = transactionId,
            TrackingNumber = trackingNumber
        };

    public Order Reject() =>
        this with { Status = OrderStatus.Rejected };

    public Order WithStatus(OrderStatus status) =>
        this with { Status = status };
}
